src/memory/redis.rs

```diff
@@ -0,0 +1,253 @@
+use std::cmp::Ordering;
+use std::{sync::Arc};
+use std::error::Error;
+use redis::{Client};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tokio::runtime::Runtime;
+
+use crate::{LLM, Memory, MemoryProvider, RelevantMemory};
+
+use tokio::{sync::Mutex};
+
+use super::MemorySystem;
+
+use async_trait::async_trait;
+
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct RedisPayload {
+    content: String,
+    recall: f32,
+    recency: f32,
+}
+
+pub struct RedisMemorySystem {
+    client: redis::Client,
+    latest_point_id: Arc<Mutex<u64>>,
+    index_name: String,
+}
+
+#[async_trait]
+impl MemorySystem for RedisMemorySystem {
+    async fn store_memory(&mut self, llm: &LLM, memory: &str) -> Result<(), Box<dyn Error>> {
+        let embedding = llm.model.get_base_embed(memory).await?;
+        let mut con = self.client.get_tokio_connection().await?;
+
+        let memory_struct = RedisPayload {
+            content: memory.to_string(),
+            recency: 1.,
+            recall: 1.,
+        };
+
+        let mut latest_point_id = self.latest_point_id.lock().await;
+        *latest_point_id += 1;
+        let point_id = latest_point_id.to_string();
+
+        let memory_json = serde_json::to_value(&memory_struct)?;
+        let embedding_json = serde_json::to_value(&embedding)?;
+
+        redis::cmd("JSON.SET")
+            .arg(&point_id)
+            .arg("$")
+            .arg(format!(r#"{{"memory": {}, "embedding": {}}}"#, memory_json, embedding_json))
```

Comment:
It would probably be best to define some sort of
```rust
#[derive(Serialize, Deserialize)]
pub struct EmbeddedMemory {
    memory: String,
    embedding: String
}
```
and then use `.arg(serde_json::to_string(&embedded_memory))`